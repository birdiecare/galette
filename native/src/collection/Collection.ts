import ListView from "deprecated-react-native-listview";
import CoreCollection from "@galette/core/dist/store/Collection";

export default class Collection extends CoreCollection
{
  dataSource() {
    let ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2})
    let items = this.items();

    return ds.cloneWithRows(items);
  }
}
