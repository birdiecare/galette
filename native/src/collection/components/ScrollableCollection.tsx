import React, { Component, ReactNode } from 'react'
import {RefreshControl, ScrollView, StyleProp, ViewStyle} from "react-native";
import CollectionComponent, { Props as CollectionComponentProps } from "./CollectionComponent";
import InfiniteScrollView from "../../infinite-scroll-view/InfiniteScrollView";
import Collection from "../Collection";

type Props = CollectionComponentProps & {
    collection: Collection;
    onRefresh: (page?: number) => void;
    header?: ReactNode;
    numberOfItemsForFullPage?: StyleProp<ViewStyle>;
    style?: any;
}

export default class ScrollableCollection extends Component<Props, {}>
{
    static defaultProps = {
        numberOfItemsForFullPage: 10,
        header: null,
    };

    componentWillMount() {
        this.props.onRefresh();
    }

    render() {
        let { collection } = this.props;

        return (
          <React.Fragment>
            <InfiniteScrollView
                style={this.props.style || {flex: 1}}
                refreshControl={
                    <RefreshControl
                        refreshing={collection.isLoading()}
                        onRefresh={() => this.props.onRefresh(1)}
                        title="Loading..."
                        tintColor="#fff"
                        titleColor="#fff" />
                }
                isLoading={collection.isLoading()}
                canLoadMore={collection.hasMore()}
                displayLoading={this._shouldDisplayBottomLoading(collection)}
                distanceToLoadMore={100}
                onLoadMoreAsync={() => {
                    this.props.onRefresh(collection.getPage() + 1);
                }}>

                {this.props.header}

                <CollectionComponent {...this.props} />
            </InfiniteScrollView>
          </React.Fragment>
        )
    }

    _shouldDisplayBottomLoading(collection) {
        return collection.getPage() > 1 || collection.items().length > this.props.numberOfItemsForFullPage;
    }
}
