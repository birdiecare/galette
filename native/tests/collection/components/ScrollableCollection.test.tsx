import React from "react";
import renderer from "react-test-renderer";
import { Text, ScrollView } from "react-native";
import ScrollableCollection from "../../../src/collection/components/ScrollableCollection";
import Collection from "../../../src/collection/Collection";

const noop = () => {};
const renderRow = () => <Text>Row</Text>;

test("it displays only one loading indicator on the first page", () => {
  const collection = new Collection({
    items: [{ foo: "bar" }],
    loading: true,
    up_to_page: 1,
    total_items: 10
  });

  const tree = renderer
    .create(
      <ScrollableCollection
        onRefresh={noop}
        renderRow={renderRow}
        collection={collection}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("it displays the bottom loading indicator when more than one page has been loaded", () => {
  const collection = new Collection({
    items: [{ foo: "bar" }, { foo: "baz" }],
    loading: true,
    up_to_page: 2,
    total_items: 10
  });

  const tree = renderer
    .create(
      <ScrollableCollection
        onRefresh={noop}
        renderRow={renderRow}
        collection={collection}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it("passes a testID prop down to the ScrollView component", () => {
  const collection = new Collection({
    items: [{ foo: "bar" }],
    loading: true,
    up_to_page: 1,
    total_items: 10
  });

  const tree = renderer.create(
    <ScrollableCollection
      testID="test-id"
      onRefresh={noop}
      renderRow={renderRow}
      collection={collection}
    />
  );

  const scrollView = tree.root.findByType(ScrollView);
  expect(scrollView.props.testID).toBe("test-id");
});
