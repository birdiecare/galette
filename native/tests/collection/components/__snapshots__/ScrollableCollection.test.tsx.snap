// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it displays only one loading indicator on the first page 1`] = `
<RCTScrollView
  canLoadMore={true}
  displayLoading={false}
  distanceToLoadMore={100}
  isLoading={true}
  onLoadMoreAsync={[Function]}
  onScroll={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
      tintColor="#fff"
      title="Loading..."
      titleColor="#fff"
    />
  }
  renderLoadingErrorIndicator={[Function]}
  renderLoadingIndicator={[Function]}
  scrollEventThrottle={100}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTRefreshControl />
  <View>
    <RCTScrollView
      dataSource={
        ListViewDataSource {
          "items": 1,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
      style={undefined}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Row
        </Text>
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;

exports[`it displays the bottom loading indicator when more than one page has been loaded 1`] = `
<RCTScrollView
  canLoadMore={true}
  displayLoading={true}
  distanceToLoadMore={100}
  isLoading={true}
  onLoadMoreAsync={[Function]}
  onScroll={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
      tintColor="#fff"
      title="Loading..."
      titleColor="#fff"
    />
  }
  renderLoadingErrorIndicator={[Function]}
  renderLoadingIndicator={[Function]}
  scrollEventThrottle={100}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTRefreshControl />
  <View>
    <RCTScrollView
      dataSource={
        ListViewDataSource {
          "items": 2,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
      style={undefined}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Row
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Row
        </Text>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "justifyContent": "center",
          "padding": 20,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
      />
    </View>
  </View>
</RCTScrollView>
`;
